File: tools/common/VMRestSso.Common/RestSsoAdminSnapIn.Service/OidcClientResource.cs
/*
 * Copyright © 2012-2015 VMware, Inc.  All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the “License”); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS, without
 * warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */


using System;
using System.Collections.Generic;
using System.Net;
using Vmware.Tools.RestSsoAdminSnapIn.Core.Serialization;
using Vmware.Tools.RestSsoAdminSnapIn.Core.Web;
using Vmware.Tools.RestSsoAdminSnapIn.Dto;

namespace Vmware.Tools.RestSsoAdminSnapIn.Service.OidcClient
{
    public class OidcClientService
    {
        private readonly IWebRequestManager _webRequestManager;
        private readonly IServiceConfigManager _serviceConfigManager;
        public OidcClientService(IWebRequestManager webRequestManager, IServiceConfigManager serviceConfigManager)
        {
            _webRequestManager = webRequestManager;
            _serviceConfigManager = serviceConfigManager;
        }

        public bool Delete(ServerDto serverDto, string tenant, OidcClientDto oidcClientDto, Token token)
        {
            tenant = Uri.EscapeDataString(tenant);
            var  clientId = Uri.EscapeDataString(oidcClientDto.ClientId);
            var url = string.Format(_serviceConfigManager.GetOidcClientEndPoint(), serverDto.Protocol, serverDto.ServerName, serverDto.Port, tenant, clientId);
            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
            var requestConfig = new RequestSettings
            {
                Method = HttpMethod.Delete,
            };
            var headers = ServiceHelper.AddHeaders(ServiceConstants.JsonContentType);
            var json = "access_token=" + token.AccessToken + "&token_type=" + token.TokenType.ToString().ToLower();
            var result = _webRequestManager.GetResponse(url, requestConfig, headers, null, json);
            return string.IsNullOrEmpty(result);
        }

        public OidcClientDto Create(ServerDto serverDto, string tenant, OidcClientMetadataDto oidcClientMetadataDto, Token token)
        {
            tenant = Uri.EscapeDataString(tenant);
            var json = JsonConvert.Serialize(oidcClientMetadataDto);
            var url = string.Format(_serviceConfigManager.GetOidcClientsEndPoint(), serverDto.Protocol, serverDto.ServerName, serverDto.Port, tenant);
            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
            var requestConfig = new RequestSettings
            {
                Method = HttpMethod.Post,
            };
            var headers = ServiceHelper.AddHeaders(ServiceConstants.JsonContentType);
            json = "access_token=" + token.AccessToken + "&token_type=" + token.TokenType.ToString().ToLower() + "&" + json;
            var response = _webRequestManager.GetResponse(url, requestConfig, headers, null, json);
            return JsonConvert.Deserialize<OidcClientDto>(response);
        }
        public OidcClientDto Get(ServerDto serverDto, string tenant, OidcClientDto oidcClientDto, Token token)
        {
            tenant = Uri.EscapeDataString(tenant);
            var clientId = Uri.EscapeDataString(oidcClientDto.ClientId);
            var url = string.Format(_serviceConfigManager.GetOidcClientPostEndPoint(), serverDto.Protocol, serverDto.ServerName, serverDto.Port, tenant, clientId);
            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
            var requestConfig = new RequestSettings
            {
                Method = HttpMethod.Post,
            };
            var headers = ServiceHelper.AddHeaders(ServiceConstants.JsonContentType);
            var json = "access_token=" + token.AccessToken + "&token_type=" + token.TokenType.ToString().ToLower();
            var response = _webRequestManager.GetResponse(url, requestConfig, headers, null, json);
            return JsonConvert.Deserialize<OidcClientDto>(response);
        }
        public List<OidcClientDto> GetAll(ServerDto serverDto, string tenant, Token token)
        {
            tenant = Uri.EscapeDataString(tenant);
            var url = string.Format(_serviceConfigManager.GetOidcClientsPostEndPoint(), serverDto.Protocol, serverDto.ServerName, serverDto.Port, tenant);
            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
            var requestConfig = new RequestSettings
            {
                Method = HttpMethod.Post,
            };
            var headers = ServiceHelper.AddHeaders(ServiceConstants.JsonContentType);
            var json = "access_token=" + token.AccessToken + "&token_type=" + token.TokenType.ToString().ToLower();
            var response = _webRequestManager.GetResponse(url, requestConfig, headers, null, json);
            return JsonConvert.Deserialize <List<OidcClientDto>>(response);
        }

        public OidcClientDto Update(ServerDto serverDto, string tenant, string clientId, OidcClientMetadataDto oidcClientDto, Token token)
        {
            tenant = Uri.EscapeDataString(tenant);
            clientId = Uri.EscapeDataString(clientId);
            var json = JsonConvert.Serialize(oidcClientDto);
            var url = string.Format(_serviceConfigManager.GetOidcClientEndPoint(), serverDto.Protocol, serverDto.ServerName, serverDto.Port, tenant, clientId);
            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
            var requestConfig = new RequestSettings
            {
                Method = HttpMethod.Put,
            };
            var headers = ServiceHelper.AddHeaders(ServiceConstants.JsonContentType);
            json = "access_token=" + token.AccessToken + "&token_type=" + token.TokenType.ToString().ToLower() + "&" + json;
            var response = _webRequestManager.GetResponse(url, requestConfig, headers, null, json);
            return JsonConvert.Deserialize<OidcClientDto>(response);
        }
    }
}

File: vmidentity/rest/idm/client/src/main/java/com/vmware/identity/rest/idm/client/OidcClientResource.java
/*
 *  Copyright (c) 2012-2015 VMware, Inc.  All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
 *  use this file except in compliance with the License.  You may obtain a copy
 *  of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, without
 *  warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 *  License for the specific language governing permissions and limitations
 *  under the License.
 */
package com.vmware.identity.rest.idm.client;

import static com.vmware.identity.rest.core.client.RequestExecutor.execute;
import static com.vmware.identity.rest.core.client.RequestExecutor.executeAndReturnList;
import static com.vmware.identity.rest.core.client.URIFactory.buildURI;

import java.io.IOException;
import java.net.URI;
import java.util.List;

import org.apache.http.HttpException;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpPut;

import com.vmware.identity.rest.core.client.BaseClient;
import com.vmware.identity.rest.core.client.ClientResource;
import com.vmware.identity.rest.core.client.RequestFactory;
import com.vmware.identity.rest.core.client.exceptions.ClientException;
import com.vmware.identity.rest.core.client.exceptions.WebApplicationException;
import com.vmware.identity.rest.core.client.methods.HttpDeleteWithBody;
import com.vmware.identity.rest.idm.data.OIDCClientDTO;
import com.vmware.identity.rest.idm.data.OIDCClientMetadataDTO;

/**
 * The {@code OidcClientResource} is effectively a container that gathers all of
 * the commands related to the Open ID Connect (OIDC) clients for a tenant. These commands
 * are all gathered under the {@value #OIDC_CLIENT_URI} endpoint on the REST server.
 */
public class OidcClientResource extends ClientResource {

    private static final String OIDC_CLIENT_URI = "/idm/tenant/%s/oidcclient";
    private static final String OIDC_CLIENT_ID_URI = OIDC_CLIENT_URI + "/%s";

    private static final String OIDC_CLIENT_POST_URI = "/idm/post/tenant/%s/oidcclient";
    private static final String OIDC_CLIENT_ID_POST_URI = OIDC_CLIENT_POST_URI + "/%s";

    public OidcClientResource(BaseClient parent) {
        super(parent);
    }

    /**
     * Register a new Open ID Connect client.
     *
     * <p><b>Required Role:</b> {@code TrustedUser}.
     *
     * @param tenant the name of the tenant to register the client with.
     * @param metadata the metadata necessary for registering the client.
     * @return the newly registered client.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public OIDCClientDTO register(String tenant, OIDCClientMetadataDTO metadata) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), OIDC_CLIENT_URI, tenant);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken(), metadata);
        return execute(parent.getClient(), post, OIDCClientDTO.class);
    }

    /**
     * Request the list of Open ID Connect clients associated with a tenant.
     *
     * <p><b>Required Role:</b> {@code TrustedUser}.
     *
     * @param tenant the name of the tenant to request the clients from.
     * @return a list of OIDC clients.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public List<OIDCClientDTO> getAll(String tenant) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), OIDC_CLIENT_POST_URI, tenant);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return executeAndReturnList(parent.getClient(), post, OIDCClientDTO.class);
    }

    /**
     * Request a specific Open ID Connect client.
     *
     * <p><b>Required Role:</b> {@code TrustedUser}.
     *
     * @param tenant the name of the tenant to request the client from.
     * @param clientId the identifier of the client to request.
     * @return the requested client.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public OIDCClientDTO get(String tenant, String clientId) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), OIDC_CLIENT_ID_POST_URI, tenant, clientId);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return execute(parent.getClient(), post, OIDCClientDTO.class);
    }

    /**
     * Update an Open ID Connect client.
     *
     * <p><b>Required Role:</b> {@code TrustedUser}.
     *
     * @param tenant the name of the tenant containing the client.
     * @param clientId the identifier of the client being updated.
     * @param metadata the metadata contents used to update the client.
     * @return the newly updated client.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public OIDCClientDTO update(String tenant, String clientId, OIDCClientMetadataDTO metadata) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), OIDC_CLIENT_ID_URI, tenant, clientId);

        HttpPut put = RequestFactory.createPutRequest(uri, parent.getToken(), metadata);
        return execute(parent.getClient(), put, OIDCClientDTO.class);
    }

    /**
     * Delete an Open ID Connect client.
     *
     * <p><b>Required Role:</b> {@code TrustedUser}.
     *
     * @param tenant the name of the tenant containing the client.
     * @param clientId the identifier of the client being deleted.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public void delete(String tenant, String clientId) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), OIDC_CLIENT_ID_URI, tenant, clientId);

        HttpDeleteWithBody delete = RequestFactory.createDeleteRequest(uri, parent.getToken());
        execute(parent.getClient(), delete);
    }

}
